src/parser/jobs/drg/modules/Positionals.js

```diff
@@ -0,0 +1,222 @@
+	static title = t('drg.positionals.title')`Positionals`
+	static dependencies = [
+		'combatants',
+		'suggestions',
+	]
+
+	// tracking raiden thrust procs
+	_rtCombos = []
+	_currentCombo = null
+
+	// true north tracking
+	_tnCharges = 2
+	_nextChargeIn = null
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('cast', {by: 'player', abilityId: ROTATION_IDS}, this._onGCD)
+		this.addHook('cast', {by: 'player', abilityId: PROC_IDS}, this._onProcGCD)
+		this.addHook('applybuff', {by: 'player', abilityId: STATUSES.RAIDEN_THRUST_READY.id}, this._procSuccess)
+		this.addHook('applybuff', {by: 'player', abilityId: STATUSES.TRUE_NORTH.id}, this._tnUsed)
+		this.addHook('complete', this._onComplete)
+	}
+
+	// duplicate code - eventually may want in single location
+	createTimelineButton(timestamp) {
+		return <Button
+			circular
+			compact
+			icon="time"
+			size="small"
+			onClick={() => this.timeline.show(timestamp - this.parser.fight.start_time,	timestamp - this.parser.fight.start_time)}
+			content={this.parser.formatTimestamp(timestamp)}
+		/>
+	}
+	// end duplicate code
+
+	_onGCD(event) {
+		if (this._currentCombo) {
+			// did we use the proc? we might not due to invuln windows
+			if (event.ability.guid === ACTIONS.RAIDEN_THRUST.id) {
+				this._rtCombos.used = true
+			}
+
+			// close the window
+			this._rtCombos.push(this._currentCombo)
+			this._currentCombo = null
+		}
+	}
+
+	_onProcGCD(event) {
+		// get action data
+		if (!this._currentCombo) {
+			// if we don't have an open combo window, open one
+			this._currentCombo = {
+				next: NEXT_COMBO[event.ability.guid],
+				success: false,
+				used: false,
+				trueNorthCharges: this._tnCharges,
+				trueNorthCast1: this.combatants.selected.hasStatus(STATUSES.TRUE_NORTH.id),
+				time: event.timestamp,
+			}
+		} else if (this._currentCombo) {
+			// if the action isn't the next expected one, close the current window (not expecting a proc
+			// due to broken combo).
+			if (this._currentCombo.next !== event.ability.guid) {
+				// discard, let the broken combo module yell about this
+				this._currentCombo = null
+			} else {
+				// add the time
+				this._currentCombo.time = event.timestamp
+
+				// check for true north still active
+				this._currentCombo.trueNorthCast2 = this.combatants.selected.hasStatus(STATUSES.TRUE_NORTH.id)
+			}
+		}
+	}
+
+	// called exactly when the timestamp resolves
+	_updateTnCharges() {
+		this._tnCharges += 1
+
+		if (this._tnCharges > TRUE_NORTH_CHARGES) {
+			// clamp to max
+			this._tnCharges = TRUE_NORTH_CHARGES
+		}
+
+		if (this._tnCharges < TRUE_NORTH_CHARGES) {
+			// if we're below the max, queue another charge since the cd will keep ticking
+			this.addTimestampHook(this.parser.currentTimestamp + TRUE_NORTH_CD, this._updateTnCharges)
+		}
+	}
+
+	_tnUsed(event) {
+		// remove charge
+		this._tnCharges -= 1
+
+		// if we're below 0, clamp
+		if (this._tnCharges < 0) {
+			this._tnCharges = 0
+		}
```

Comment:
If you want to shorten this, it can all be simplified down to `this._tnCharges = Math.max(this._tnCharges - 1, 0)`. That said, it might instead be worth throwing a broken log error if they somehow hit negative charges, because that should be physically impossible. Up to you really.

Reply:
I'd be ok with throwing/showing an error. Is there a standard way to do that here, or should I just set a flag and display a message in the output section?

Reply:
You can depend on the [`brokenLog`](https://github.com/xivanalysis/xivanalysis/blob/master/src/parser/core/modules/BrokenLog.tsx) module, and use it's API to flag an issue. The rest is handled by core.